src/flint/types/fq_default.pyx

```diff
@@ -0,0 +1,886 @@
+from flint.pyflint cimport global_random_state
+from flint.types.fmpz cimport fmpz, any_as_fmpz
+from flint.types.nmod cimport nmod
+from flint.types.fmpz_mod cimport fmpz_mod
+from flint.types.fmpz_poly cimport fmpz_poly, any_as_fmpz_poly, fmpz_poly_set_list
+from flint.types.fmpz_mod_poly cimport fmpz_mod_poly, fmpz_mod_poly_ctx
+from flint.types.nmod_poly cimport nmod_poly
+from flint.utils.typecheck cimport typecheck
+
+cdef class fq_default_ctx:
+    r"""
+    Context object for creating :class:`~.fq_default`.
+
+    Finite fields can be initialized in one of two possible ways. The
+    first is by providing characteristic and degree:
+
+        >>> fq_default_ctx(5, 2, 'y', fq_type=1)
+        fq_default_ctx(5, 2, 'y', x^2 + 4*x + 2, 'FQ_ZECH')
+
+    The second is by giving an irreducible polynomial of type
+    :class:`~.nmod_poly` or :class:`~.fmpz_mod_poly`:
+
+        >>> from flint import fmpz_mod_poly_ctx
+        >>> mod = fmpz_mod_poly_ctx(11)([1,0,1])
+        >>> fq_default_ctx(modulus=mod, fq_type=2)
+        fq_default_ctx(11, 2, 'x', x^2 + 1, 'FQ_NMOD')
+
+    For more details, see the documentation of :method:`~.set_from_order`
+    and :method:`~.set_from_modulus`.
+    """
+    def __cinit__(self):
+        pass
+
+    def __dealloc__(self):
+        if self._initialized:
+            fq_default_ctx_clear(self.val)
```

Comment:
There was some discussion about this in the mpoly PR so I need to find out what the situation is. Apparently Cython does not guarantee that `__dealloc__` is only ever called once. If that is the case then it means that we also need to set `self._initialized = False` here after calling `clear` to prevent a double-free.